File: backend/src/modules/users/infrastructure/persistence/user.mapper.ts
import { User, type UserProps } from '@modules/users/domain/entities/user.entity';
import { KEY_PREFIXES, SK_VALUES } from '@shared/infrastructure/dynamodb/single-table.constants';

export interface UserItem {
  PK: string;
  SK: string;
  GSI1PK: string;
  GSI1SK: string;
  id: string;
  email: string;
  name: string;
  passwordHash: string;
  createdAt: string;
}

export class UserMapper {
  static toItem(user: User): UserItem {
    return {
      PK: `${KEY_PREFIXES.USER}${user.id}`,
      SK: SK_VALUES.PROFILE,
      GSI1PK: `${KEY_PREFIXES.EMAIL}${user.email.toLowerCase()}`,
      GSI1SK: SK_VALUES.USER,
      id: user.id,
      email: user.email,
      name: user.name,
      passwordHash: user.passwordHash,
      createdAt: user.createdAt,
    };
  }

  static toDomain(item: Record<string, unknown>): User {
    const props: UserProps = {
      id: item.id as string,
      email: item.email as string,
      name: item.name as string,
      passwordHash: item.passwordHash as string,
      createdAt: item.createdAt as string,
    };
    return new User(props);
  }
}

File: backend-v2/crates/infrastructure/src/dynamodb/user/user_mapper.rs
use std::collections::HashMap;

use aws_sdk_dynamodb::types::AttributeValue;
use chrono::{DateTime, Utc};
use domain::modules::user::{entity::User, repository::RepositoryError};

use crate::dynamodb::{keys_table::UserKey, prefixes::EMAIL_PREFIX};

pub struct UserMapper;

impl UserMapper {
    pub fn to_item(user: &User) -> HashMap<String, AttributeValue> {
        HashMap::from([
            ("PK".into(), AttributeValue::S(UserKey::pk(user.id()))),
            ("SK".into(), AttributeValue::S(UserKey::sk(None))),
            (
                "GSI1PK".into(),
                AttributeValue::S(format!(
                    "{}{}",
                    EMAIL_PREFIX,
                    user.email().to_lowercase().trim()
                )),
            ),
            ("GSI1SK".into(), AttributeValue::S(UserKey::gsi1sk(None))),
            ("id".into(), AttributeValue::S(user.id().to_string())),
            ("name".into(), AttributeValue::S(user.name().to_string())),
            ("email".into(), AttributeValue::S(user.email().to_string())),
            (
                "passwordHash".into(),
                AttributeValue::S(user.password_hash().to_string()),
            ),
            (
                "createAt".into(),
                AttributeValue::S(user.create_at().to_rfc3339()),
            ),
        ])
    }

    pub fn to_domain(item: &HashMap<String, AttributeValue>) -> Result<User, RepositoryError> {
        Ok(User {
            id: string_field(item, "id")?,
            email: string_field(item, "email")?,
            name: string_field(item, "name")?,
            create_at: datetime_field(item, "createAt")?,
            password_hash: string_field(item, "passwordHash")?,
        })
    }
}

fn string_field(
    item: &HashMap<String, AttributeValue>,
    key: &str,
) -> Result<String, RepositoryError> {
    item.get(key)
        .and_then(|v| v.as_s().ok())
        .map(|s| s.to_string())
        .ok_or_else(|| RepositoryError::DatabaseError(format!("falta el atributo '{key}'")))
}

fn datetime_field(
    item: &HashMap<String, AttributeValue>,
    key: &str,
) -> Result<DateTime<Utc>, RepositoryError> {
    let value = string_field(item, key)?;

    DateTime::parse_from_rfc3339(&value)
        .map(|dt| dt.with_timezone(&Utc))
        .map_err(|e| {
            RepositoryError::DatabaseError(format!(
                "el atributo '{key}' no contiene un timestamp válido: {e}"
            ))
        })
}
